### ISSUE TITLE
Non deterministic output for recursive type var default

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ty#3804]
### Summary

Codex root caused the `steam.py` non determinism to

`commands.py`:

```py
from typing import TYPE_CHECKING, Generic
from typing_extensions import TypeVar

if TYPE_CHECKING:
    from cog import Bot

CogT = TypeVar("CogT", bound="Bot", default="Bot", covariant=True)


class GroupMixin(Generic[CogT]):
    def method(self): ...


def call_method(value: object):
    if isinstance(value, GroupMixin):
        value.method()
```

`cog.py`:

```py
from commands import GroupMixin


class Group(GroupMixin): ...


class Bot(GroupMixin): ...
```

[Playground](https://play.ty.dev/791279f8-25ed-4de8-8fa1-4a4cf708b0f1)

* `ty check commands.py`: pass
* `ty check commands.py cog.py`: Arguments to `value.method` in `commands.py` are incorrect (even when using `RAYON_NUM_THREADS=1`)

**Comment by MichaReiser** [comment]
Uhh, the issue seems to be that `bound_typevar_default_type_cycle_recover` always returns `None`... Which means we never fixpoint iterate. @AlexWaygood is this intentional?

**Comment by AlexWaygood** [comment]
> [@AlexWaygood](https://github.com/AlexWaygood) is this intentional?

I assume from this question that I wrote this function? Probably not intentional 😆

**Comment by MichaReiser** [comment]
Ah sorry, you only moved it. That's what you get for touching all ty code 😆

**Comment by MichaReiser** [comment]
Okay, this was introduced in https://github.com/astral-sh/ruff/pull/23652/files by @charliermarsh 

Charlie also already tried fixing this in https://github.com/astral-sh/ruff/commit/3d3841454ad03edd496d522985f76ad17430a8da

**Comment by charliermarsh** [comment]
You can assign to me if needed (though can't get to it today).

**Comment by MichaReiser** [comment]
I give it a try with codex. The issue basically is. Proper cycle recovery reintroduces the stack overflow. I now have codex trying to fix the stack overflow instead

**Comment by snk0911** [comment]
I have a question, and sorry in advance crashing your conversation about the bug fix. 

Do you actually all use AI to fix stuff here? I am just curious how eerything works since I want to help out, too. But it is hard to get into the topic :)

**Comment by MichaReiser** [comment]
We heavily use AI but everyone of us is still responsible for their contribution. That means, we use AI to root cause issues (root causing this would have taken me hours but took codex minutes) and have it come up with an initial fix. I review every fix and often iterate heavily before opening a PR.

**Comment by snk0911** [comment]
@MichaReiser Thank you for giving me insides :) I was exactly thinking about root causing. This is something I have trouble with often but when it comes to code and logical thinking, iterating over the initial fix is still mandatory.
